_posts/2019/2019-10-10-building-a-drawer-component-in-react.md

```diff
@@ -0,0 +1,805 @@
+
+Now that we've fixed our button's accesibility issues we can focus back on our goal of showing and hiding content in the Drawer. Since the height of the element starts out as 0 we can simply have its height change to a larger value when the `is-expanded` class is active via the `this.state.active` variable. We want this value to be greater than the containing content's height could ever be. I choose `40vh` here as it is 40% of the current viewports height (and it seemed a reasonable starting place).
+
+``` scss
+.Drawer {
+  &-title {
+    ...
+  }
+
+  &-content {
+    ...
+  }
+
+  &.is-expanded {
+    .Drawer-content {
+      height: 40vh;
+    }
+  }
+}
+```
+
+Making this change now enables the element to show/hide its content. We can make
+this even more polished by adding an animated transition to the content.
+
+``` scss
+.Drawer {
+  &-title {
+    ...
+  }
+
+  &-content {
+    height: 0;
+    overflow: hidden;
+    transition: height .5s ease;
+  }
+
+  &.is-expanded {
+    .Drawer-content {
+      height: 10vh;
+    }
+  }
+}
+```
+
+Now that our `Drawer-content` section animates its height our component is starting
+to show off its basic functionality. With a couple more tweaks to the styles below, we've got our first iteration complete.
+
+``` scss
+.Drawer {
+  width: 300px; // Set parent width
+  display: inline-block; // center the element
+
+  &-title {
+    padding: 16px;
+    border: 1px solid;
+    background: #efefef;
+    cursor: pointer;
+    width: 100%; // Make title button take up full width of .Drawer
+  }
+
+  &-content {
+    background: #ddd; // Show a visible background
+    border-left: 1px solid; // Add borders to all sides except top
+    border-right: 1px solid;
+    border-bottom: 1px solid;
+    height: 0;
+    overflow: hidden;
+    transition: height .5s ease;
+  }
+
+  &.is-expanded {
+    .Drawer-content {
+      height: 40vh;
+    }
+  }
+}
+```
+
+![First iteration on Component](/img/2019/first-iteration-component.gif)
+
+## Static height and variable containers
+
+While our component now functions well enough to show/hide, there is a glaring issue with it. If you look back to our `.is-expanded` class we had to specify a static height of `40vh`. This means that our container (which can contain a dynamic amount of content) will never have a height larger than 40% of the viewport. What if the content is larger than 40%? A lot smaller? Well the results aren't great:
+
+![Static height and variable containers example](/img/2019/set-height-bug.png)
+
+That's a lot of extra space for a single list item!
+
+So how can we fix this? Well, unfortunately if you're looking for a pure CSS solution here the short answer is, you can't. Or at least all the articles I've read online claim there isn't a process and I haven't been able to figure one out either.
+
+Really the only way to determine dynamic height for the content is to calculate it using JavaScript. So let's do just that. Here's what we'll do:
```

Comment:
Really the only way to determine dynamic height for the content is to calculate it using JavaScript. Basically, we'll be calculating the space on the fly and adjusting as needed. So let's do just that. Here's what we'll do: